src/components/DropdownMenu/DropdownMenuItem.tsx

```diff
@@ -1,27 +1,126 @@
+    closeMenu,
+    children,
+    path,
+    ...props
+}: DropdownMenuItemProps<T>) => {
     const {toggle, data} = useContext(DropdownMenuContext);
-    const onClick: MouseEventHandler<HTMLElement> = useCallback(
-        (event) => {
+    const menuItemRef = useRef(null);
+
+    const {hasSubmenu, isSubmenuOpen, closeSubmenu, openSubmenu} = useSubmenu({
+        items: subMenuItems,
+        path,
+    });
+
+    const handleCloseMenu = useCallback(() => {
+        const close = () => {
+            if (closeMenu) {
+                closeMenu();
+            } else {
+                toggle(false);
+            }
+        };
+
+        if (hasSubmenu) {
+            closeSubmenu();
+            // Wait for submenu to close
+            requestAnimationFrame(close);
+        } else {
+            close();
+        }
+    }, [closeMenu, closeSubmenu, hasSubmenu, toggle]);
+
+    const handleMenuItemClick = useCallback(
+        (event: React.MouseEvent<HTMLElement, MouseEvent>) => {
+            if (hasSubmenu) {
+                return;
+            }
+
             action?.(event, data as unknown as T);
-            toggle(false);
+            handleCloseMenu();
         },
-        [action, data, toggle],
+        [action, data, handleCloseMenu, hasSubmenu],
     );
 
+    const extraProps = useMemo(() => {
+        return {
+            ...props.extraProps,
+            onMouseEnter: (
+                event: React.MouseEvent<HTMLDivElement, MouseEvent> &
+                    React.MouseEvent<HTMLAnchorElement, MouseEvent>,
+            ) => {
+                props.extraProps?.onMouseEnter?.(event);
+                if (hasSubmenu) {
+                    openSubmenu();
+                }
+            },
+            onMouseLeave: (
+                event: React.MouseEvent<HTMLDivElement, MouseEvent> &
+                    React.MouseEvent<HTMLAnchorElement, MouseEvent>,
+            ) => {
+                props.extraProps?.onMouseLeave?.(event);
+                if (hasSubmenu) {
+                    closeSubmenu();
+                }
+            },
+        };
+    }, [props.extraProps, closeSubmenu, hasSubmenu, openSubmenu]);
+
     return (
-        <Menu.Item onClick={onClick} {...props}>
-            {text || children}
-        </Menu.Item>
+        <>
+            <Menu.Item
+                ref={menuItemRef}
+                {...props}
+                extraProps={extraProps}
+                onClick={handleMenuItemClick}
+            >
+                <div className={cnDropdownMenu('menu-item-content')}>
+                    <div>{text || children}</div>
+                    {hasSubmenu && (
+                        <Icon
+                            data={Chevron}
+                            className={cnDropdownMenu('menu-item-chevron')}
+                            size={8}
+                        />
+                    )}
+                </div>
+            </Menu.Item>
+            {hasSubmenu && subMenuItems && (
```

Comment:
placing submenu here doesn't looks right. `Popup` require just link for element to position against, so it could be placed outside of DropdownMenuItem

Reply:
Also, it won't be needed to store refs to menu items in the popup and it will be possible to render submenu in custom menu